### ISSUE TITLE
1:1 pod to node assignment

### ISSUE THREAD

**Issue by ahg-g** [kubernetes/kubernetes#105287]
#### What would you like to be added:

_Summary:_
An explicit API and scheduler support 1:1 pod to node assignment. 

_Proposal_
The idea is to create a scheduler filter plugin such that given a pod and a node, a node is filtered out if any of the following is true (assume the API is the annotation `k8s.io/pod-per-node=true`, more on that below):
1. The incoming pod is not a daemonset pod (by checking OwnerReference) and there is a pod with `k8s.io/pod-per-node=true` already assigned to the node, 
2. If the incoming pod has the `k8s.io/pod-per-node=true` annotation and the node hosts at least one non daemonset pod.

The above logic would also be added to the set of predicates that kubelet tests on pod admission. This will prevent users from working around the scheduler to avoid 1:1 enforcement.

_The API_

An annotation is not a proper API, so another point of discussion is what the API should look like in the pod spec? One suggestion is to add an enum: `NodeAssignmentPolicy` or `AssignmentPolicy`. Should we have this inside a struct? I wish we had a SchedulingPolicy struct in which we put also PodTopologySpread, NodeAffinity, PodAffinity etc. parameters.

#### Why is this needed:

Many HPC and batch workloads desire 1:1 pod to node assignment. Currently there are two approaches:
1) set resource requests to something high enough to prevent scheduling any other pods but daemonset pods. This is tricky to do and error prone, especially for smaller nodes.
2) Use pod anti affinity. Users will need to define the same set of labels on all pods that needs to be repelled, whether or not they actually request 1:1 pod to node assignment, it is also expensive to calculate especially  for cluster autoscaler.

/sig scheduling

**Comment by Huang-Wei** [comment]
I'm thinking about achieving this without the intervening of scheduler.

Can the user achieve this by writing a controller/operator like this?

1. watch the assigned/scheduled pods (with non-nil .spec.nodeName) with the annotation
1. if an assigned pod gets created, taint the node with a user-defined taint {foo/bar:NoSchedule}
1. if an assigned pod gets deleted, remove the reserved taint off the node

And for sure, to ensure regular DasemonSet to be scheduled, the DaemonSets should either carry a wildcard (`{operator: Exists}`) or mapped toleration.

> I wish we had a SchedulingPolicy struct in which we put also PodTopologySpread, NodeAffinity, PodAffinity etc. parameters.

I had the same idea when I built PodTopologySpread, sth. like `SchedulingClass`. But I guess it's too late to introduce this while also keeping things backward-compatible.

**Comment by alculquicondor** [comment]
> how to prevent other pods scheduled to this node

You should still taint your nodes in that case. The idea of this feature is that, once a pod with `pod-per-node` lands on a node, no other pod can be scheduled there (that is not a daemonset).

> if we want to release this nodes for others pods, what we want to do.

You remove the pod with `pod-per-node`.

**Comment by kerthcet** [comment]
thanks for the details.

**Comment by kerthcet** [comment]
Hi all, I'm interested with this feature, I can work on this if necessary.

**Comment by alculquicondor** [comment]
Hi Kante, thank you for the interest, but this is already assigned.

**Comment by ArangoGutierrez** [comment]
++

**Comment by ahg-g** [comment]
@ahmad-diaa can you please submit a KEP as soon as possible, we need to get consensus on the API, which may take sometime.

**Comment by ahg-g** [comment]
/cc @x13n 

for CA input

**Comment by x13n** [comment]
I don't see any issues from CA point of view. We may want to optimize binpacking by making CA aware of this feature, but otherwise I'd expect imported scheduler code to just work.

**Comment by stek29** [comment]
How is this different from using podAntiAffinity and labels?

something like this:
```

**Comment by stek29** [comment]
How is this different from using podAntiAffinity and labels?
Something like this:

```yaml
kind: Pod
metadata:
  name: pod-per-node-example
  labels:
    k8s.io/pod-per-node: "true"
spec:
  affinity:
    podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: k8s.io/pod-per-node
            operator: Exists
        topologyKey: kubernetes.io/hostname
```

**Comment by alculquicondor** [comment]
Did you mean `podAntiAffinity`? The complication of the above approach is that every pod has to have such label, even if such pod doesn't need to be 1:1.

**Comment by stek29** [comment]
@alculquicondor thanks, it makes sense now.
also, I guess specific option for 1:1 pod should be more optimal for scheduler than antiAffinity on kubernetes.io/hostname, I guess.

**Comment by alculquicondor** [comment]
Yes, ideally it should be faster to calculate. There's an open KEP if you are interested in following up #3098

**Comment by sftim** [comment]
> I'm thinking about achieving this without the intervening of scheduler.
> 
> Can the user achieve this by writing a controller/operator like this?
> 
>     1. watch the assigned/scheduled pods (with non-nil .spec.nodeName) with the annotation
> 
>     2. if an assigned pod gets created, taint the node with a user-defined taint {foo/bar:NoSchedule}
> 
>     3. if an assigned pod gets deleted, remove the reserved taint off the node
> 
> 
> And for sure, to ensure regular DasemonSet to be scheduled, the DaemonSets should either carry a wildcard (`{operator: Exists}`) or mapped toleration.
> 
> > I wish we had a SchedulingPolicy struct in which we put also PodTopologySpread, NodeAffinity, PodAffinity etc. parameters.
> 
> I had the same idea when I built PodTopologySpread, sth. like `SchedulingClass`. But I guess it's too late to introduce this while also keeping things backward-compatible.

With [Pod Scheduling Readiness](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-scheduling-readiness/), I think that approach is now feasible out of tree. It might be easier to use a label not an annotation to simplify the watch.

Also, if the cluster autoscaling controller supports it, nodes could launch with the pod-per-node taint already applied, and then this out of tree controller could remove that taint provided there are no labelled Pods waiting for an empty node to launch.

That allows supporting _fresh_ node per labelled Pod, as well as node reuse.

**Comment by alculquicondor** [comment]
> there is a tainted (and drained / freshly launched / whatever) ~Pod~ Node ready for them to schedule onto

But then this means that the custom controller needs to know all the scheduling algorithms to be able to determine if a Pod fits in a Node. It can be done, but it doesn't sound elegant.

> I guess we could also use a custom resource (think GPU devices) that the 1:1 Node claims to have one of. Mutate the Pod to request exactly one of this resource.

Yes, that is a possibility as well, but it requires installing a device plugin. Doable, but we were hoping to do better.
